plugin/trino-mariadb/src/main/java/io/trino/plugin/mariadb/MariaDbClient.java

```diff
@@ -623,6 +647,102 @@ protected boolean isSupportedJoinCondition(ConnectorSession session, JdbcJoinCon
+            }
+            rowCount = max(firstNonNull(rowCount, 0L), firstNonNull(indexMaxCardinality, 0L));
+
+            TableStatistics.Builder tableStatistics = TableStatistics.builder();
+            tableStatistics.setRowCount(Estimate.of(rowCount));
+
+            // TODO statistics from ANALYZE TABLE (https://mariadb.com/kb/en/engine-independent-table-statistics/)
+            // Map<String, AnalyzeColumnStatistics> columnStatistics = statisticsDao.getColumnStatistics(table);
+            Map<String, AnalyzeColumnStatistics> columnStatistics = ImmutableMap.of();
+
+            // TODO add support for histograms https://mariadb.com/kb/en/histogram-based-statistics/
+
+            // statistics based on existing indexes
+            Map<String, ColumnIndexStatistics> columnStatisticsFromIndexes = statisticsDao.getColumnIndexStatistics(table);
+
+            if (columnStatistics.isEmpty() && columnStatisticsFromIndexes.isEmpty()) {
+                log.debug("No column and index statistics read");
+                // No more information to work on
+                return tableStatistics.build();
+            }
+
+            for (JdbcColumnHandle column : getColumns(session, table)) {
+                ColumnStatistics.Builder columnStatisticsBuilder = ColumnStatistics.builder();
+
+                String columnName = column.getColumnName();
+                AnalyzeColumnStatistics analyzeColumnStatistics = columnStatistics.get(columnName);
+                if (analyzeColumnStatistics != null) {
+                    log.debug("Reading column statistics for %s, %s from analayze's column statistics: %s", table, columnName, analyzeColumnStatistics);
+                    columnStatisticsBuilder.setNullsFraction(Estimate.of(analyzeColumnStatistics.nullsRatio()));
+                }
+
+                ColumnIndexStatistics columnIndexStatistics = columnStatisticsFromIndexes.get(columnName);
+                if (columnIndexStatistics != null) {
+                    log.debug("Reading column statistics for %s, %s from index statistics: %s", table, columnName, columnIndexStatistics);
+                    columnStatisticsBuilder.setDistinctValuesCount(Estimate.of(columnIndexStatistics.cardinality()));
+
+                    if (!columnIndexStatistics.nullable()) {
+                        double knownNullFraction = columnStatisticsBuilder.build().getNullsFraction().getValue();
+                        if (knownNullFraction > 0) {
+                            log.warn("Inconsistent statistics, null fraction for a column %s, %s, that is not nullable according to index statistics: %s", table, columnName, knownNullFraction);
+                        }
+                        columnStatisticsBuilder.setNullsFraction(Estimate.zero());
+                    }
+
+                    // row count from INFORMATION_SCHEMA.TABLES may be very inaccurate
+                    rowCount = max(rowCount, columnIndexStatistics.cardinality());
```

Comment:
the code seems to assume inaccuracy is always towards the lower values - is this a fair assumption to make?

i.e. rowCount is always lower than column index cardinality value?

Reply:
yeah, looking at a few other places the logic is that overestimating stats is better than underestimation because underestimation can actually lead to objectively bad plans.

e.g. broadcast of a table that is too large to broadcast is probably worse than a hash join even if the table could be broadcast.

Marking as resolved.